nequip/data/AtomicData.py

```diff
@@ -249,6 +250,44 @@ def from_ase(cls, atoms, r_max, **kwargs):
             **add_fields,
         )
 
+    def to_ase(self) -> Union[List[ase.Atoms], ase.Atoms]:
+        """Build a (list of) ``ase.Atoms`` object(s) from an ``AtomicData`` object.
+
+        For each unique batch number associated with AtomicDataDict.BATCH_KEY,
+        an ``ase.Atoms`` object is created. If AtomicDataDict.BATCH_KEY does not
+        exist in self, a single ``ase.Atoms`` object is created.
+
+        Returns:
+            A list of ``ase.Atoms`` objects if AtomicDataDict.BATCH_KEY is in self
+            and is not None. Otherwise, a single ``ase.Atoms`` object is returned.
+        """
+        positions = self.pos
+        atomic_nums = self.atomic_numbers
+        pbc = self.pbc if AtomicDataDict.PBC_KEY in self else None
+        cell = self.cell if AtomicDataDict.CELL_KEY in self else None
+        batch = self.batch if AtomicDataDict.BATCH_KEY in self else None
+
+        if batch is not None:
+            unique_batches = range(batch.max() + 1)
+            batch_atoms = []
+            for batch_idx in unique_batches:
+                mask = batch == batch_idx
+                mol = Atoms(
+                    numbers=atomic_nums[mask],
+                    positions=positions[mask],
+                    cell=cell[batch_idx] if cell is not None else None,
+                    pbc=pbc[batch_idx] if pbc is not None else None,
```

Comment:
There's a special case here I forgot to mention:

- `cell` can have shape `[3, 3]` instead of of `[n_batch, 3, 3]`, in which case it is shared across all examples
- similar for `pbc`, but with shape `[3]` vs shape `[n_batch, 3]`

Reply:
So the one case this still doesn't handle is the one where `cell` is _shared_ between all batches 

An elegant way to handle this is to use `.expand`, which makes a `Tensor` "bigger" without allocating more memory (by repeating a dimension of size 1). Something like:
```python
n_batches = batch.max() + 1
cell = cell.expand(n_batches, 3, 3) if cell is not None else None
pbc = pbc.expand(n_batches, 3) if pbc is not None else None
for batch_idx in range(n_batches):
    ...
```